File: platform/script-debugger/protocol/protocol-model-generator/src/org/jetbrains/protocolReader/ProtocolModelGenerator.java
package org.jetbrains.protocolReader;

import org.jetbrains.io.JsonReaderEx;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.jsonProtocol.ProtocolSchemaReaderImpl;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.FileSystems;
import java.nio.file.Files;

public class ProtocolModelGenerator {
  public static void main(String[] args) throws IOException {
    String outputDir = args[0];
    String schemaUrl = args[1];
    byte[] bytes;
    if (schemaUrl.startsWith("http")) {
      bytes = loadBytes(new URL(schemaUrl).openStream());
    }
    else {
      bytes = Files.readAllBytes(FileSystems.getDefault().getPath(schemaUrl));
    }
    JsonReaderEx reader = new JsonReaderEx(new String(bytes, StandardCharsets.UTF_8));
    reader.setLenient(true);
    new Generator(outputDir, args[2], args[3]).go(new ProtocolSchemaReaderImpl().parseRoot(reader));
  }

  @NotNull
  public static byte[] loadBytes(@NotNull InputStream stream) throws IOException {
    ByteArrayOutputStream buffer = new ByteArrayOutputStream(Math.max(stream.available(), 16 * 1024));
    byte[] bytes = new byte[1024 * 20];
    while (true) {
      int n = stream.read(bytes, 0, bytes.length);
      if (n <= 0) {
        break;
      }
      buffer.write(bytes, 0, n);
    }
    buffer.close();
    return buffer.toByteArray();
  }
}

File: platform/script-debugger/protocol/protocol-model-generator/src/ProtocolModelGenerator.kt
package org.jetbrains.protocolReader

public class ProtocolModelGenerator {
  default object {
    throws(javaClass<IOException>())
    public fun main(args: Array<String>) {
      val outputDir = args[0]
      val schemaUrl = args[1]
      val bytes: ByteArray
      if (schemaUrl.startsWith("http")) {
        bytes = loadBytes(URL(schemaUrl).openStream())
      }
      else {
        bytes = Files.readAllBytes(FileSystems.getDefault().getPath(schemaUrl))
      }
      val reader = JsonReaderEx(String(bytes, StandardCharsets.UTF_8))
      reader.setLenient(true)
      Generator(outputDir, args[2], args[3]).go(ProtocolSchemaReaderImpl().parseRoot(reader))
    }

    throws(javaClass<IOException>())
    public fun loadBytes(stream: InputStream): ByteArray {
      val buffer = ByteArrayOutputStream(Math.max(stream.available(), 16 * 1024))
      val bytes = ByteArray(1024 * 20)
      while (true) {
        val n = stream.read(bytes, 0, bytes.size())
        if (n <= 0) {
          break
        }
        buffer.write(bytes, 0, n)
      }
      buffer.close()
      return buffer.toByteArray()
    }
  }
}

fun main(args: Array<String>) = ProtocolModelGenerator.main(args)